apps/condo/domains/user/utils/serverSchema/auth.js

```diff
@@ -0,0 +1,410 @@
+    if (authFactors.confirmEmailToken !== undefined) throw new Error('confirmEmailToken is not supported yet')
+
+    const userType = userIdentity.userType
+
+    // if (!phone && !email) throw new Error('You must provide a phone number or email')
+    if (!userType) throw new Error('You must provide a user type')
+
+    // Get user
+    const { success, user } = await _getUser(userIdentity, authFactors)
+    if (!success) {
+        await _preventTimeBasedAttack(authFactors)
+        return { success: false }
+    }
+
+    // Verify the secret matches
+    const match = await _matchUser(user, authFactors)
+    if (!match.success) {
+        return {
+            success: false,
+            _error: match._error,
+        }
+    }
+
+    return {
+        success: true,
+        user,
+        confirmPhoneAction: match.confirmPhoneAction,
+    }
+}
+
+/**
+ *
+ * @param {{ phone?: string, email?: string, userType: 'staff' | 'resident' | 'service' }} userIdentity
+ * @param {{ confirmPhoneToken?: string, confirmEmailToken?: string, password?: string }} authFactors
+ * @return {Promise<{success: boolean}|{success: boolean, user: *}>}
+ * @private
+ */
+async function _getUser (userIdentity, authFactors) {
+    if (!userIdentity || typeof userIdentity !== 'object') throw new Error('You must provide userIdentity')
+
+    let phone = userIdentity.phone
+    let email = userIdentity.email
+    const userType = userIdentity.userType
+
+    if (!userType) throw new Error('You must provide a user type')
+
+    // NOTE: If password and phone/email were not transferred, but confirm token was,
+    // then we try to get phone/email from confirm token to understand who we are trying to verify
+    if (!authFactors?.password) {
+        const { keystone } = getSchemaCtx('User')
+
+        if (phone === undefined && authFactors?.confirmPhoneToken) {
+            const action = await ConfirmPhoneAction.getOne(keystone,
+                {
+                    token: authFactors.confirmPhoneToken,
+                },
+                'id phone'
+            )
+
+            if (action) {
+                phone = action.phone
+            }
+        }
+
+        // TODO(DOMA-9890): implement when added ConfirmEmailToken
+        // if (email === undefined && authFactors?.confirmEmailToken) {
+        //
+        // }
+    }
+
+    const where = { type: userType, deletedAt: null }
+    if (phone) {
+        where.phone = phone
+        // TODO(Alllex202): uncomment when active users phones will be verified
+        // where.isPhoneVerified = true
+    }
+    if (email) {
+        where.email = email
+        // TODO(DOMA-9890): uncomment when add ConfirmEmailToken and all integrations will be have verified emails
+        // where.isEmailVerified = true
+    }
+
+    const users = await itemsQuery('User', {
+        where,
+        first: 2,
+    })
+
+    if (!phone && !email)  {
+        return { success: false }
+    }
```

Comment:
mb do it before request?

Reply:
I added this check after the database query so that the execution time would be approximately the same for any user credential check